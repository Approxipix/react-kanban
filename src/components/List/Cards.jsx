import React, { Component, PureComponent } from 'react';
import { connect } from "react-redux/es/alternate-renderers";
import styled from 'styled-components'
import { Droppable, Draggable } from 'react-beautiful-dnd';

import Card from '../Card/Card'

const CardLIst = styled.div`
  min-height: 5rem;
  
  
  position: relative;
  padding: 0 .5rem;
  background-color: ${props => (props.isDraggingOver ? '#c1c1c1' : '#dfe1e6')}
  flex-grow: 1;
  
  overflow: auto;
`;

class InnerList extends Component {
  render() {
    const { cardsId, listId } = this.props;
    return (
      <>
        {cardsId.map((cardId, index) =>
          <Card
            isDraggingOver={this.props.isDraggingOver}

            key={index}
            index={index}
            listId={listId}
            cardId={cardId}
          />
        )}
      </>
    )
  }
}

class Cards extends Component {
  render() {
    const { listId, cardsId } = this.props;
    return (
      <Droppable
        type={'task'}
        droppableId={`${listId}`}
      >
        {(provided, { isDraggingOver }) => (
          <CardLIst
            {...provided.droppableProps}
            ref={provided.innerRef}
            isDraggingOver={isDraggingOver}
          >
            <InnerList
              isDraggingOver={isDraggingOver}
              listId={listId}
              cardsId={cardsId}
            />
            {provided.placeholder}
          </CardLIst>
        )}
      </Droppable>
    )
  }
}


export default Cards;
