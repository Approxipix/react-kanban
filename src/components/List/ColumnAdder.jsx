import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { addList } from '../../redux/rootReducer/actions';

const Wrapper = styled.div`
  width: 15rem;
  margin: 0 .5rem;
  padding: .6rem;
  background-color: #e3e3e3;
  border-radius: .2rem
`;

const Title = styled.h4`
  margin-bottom: .5rem;
`;

const Input = styled.input`
  width: 100%;
  margin-bottom: .5rem;
  padding: .6rem;
  border: 1px solid #bdc3c7;
  border-radius: 3px;
  box-shadow: none;
  font-size: 1rem;
  color: #40424b;
  line-height: 1rem;
  outline: none;
`;

const Actions = styled.div`
  display: flex;
  align-items: center;
  font-size: .9rem;
`;

const CreateButton = styled.button`
  border: 1px solid #bdc3c7;
  padding: .5rem;
  margin-right: .3rem;
  border-radius: 3px;
  box-shadow: none;
  color: #40424b;
  cursor: pointer;
`;

const CancelButton = styled.button`
  margin-left: .3rem;
  padding: 0;
  font-size: inherit;
  color: #3498db;
  cursor: pointer;
`;

const AddButton = styled.button`
  width: 15rem;
  margin: 0 .5rem;
  padding: .75rem;
  border-radius: .2rem;
  background: rgba(0, 0, 0, 0.1);
  color: #bdc3c7;
  font-size: 1rem; 
  text-align: left;
  transition: background-color .1s;
  cursor: pointer;
  &:hover {
    background: rgba(0, 0, 0, 0.2);
  }
`;

class BoardAdder extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
  }

  componentDidMount() {
    document.addEventListener('click', this.outerClick);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.outerClick);
  }

  outerClick = (e) => {
    let { target } = e;
    if (target.id === 'column-add-btn' || target.id === 'column-add-form') {
      return;
    }
    if (!!target.closest && (target.closest('#column-add-form'))) {
      return;
    }
    this.setState({ isOpened: false });
  };

  handleChange = event => {
    this.setState({
      title: event.target.value
    });
  };

  handleKeyDown = event => {
    if (event.keyCode === 27) {
      this.setState({
        isOpened: false
      });
    }
  };

  handleSubmit = event => {
    event.preventDefault();
    const { title } = this.state;
    if (!title) return;
    this.props.actions.addList({boardsIndex: this.props.boardIndex, listTitle: title});
    this.setState({
      isOpened: false,
      title: ''
    });
  };

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened
    })
  };

  render = () => {
    const { isOpened, title } = this.state;
    return isOpened ? (
      <Wrapper>
        <Title>
          New Column
        </Title>
        <form onSubmit={this.handleSubmit} id="column-add-form">
          <Input
            autoFocus
            type="text"
            placeholder="Column name"
            value={title}
            onKeyDown={this.handleKeyDown}
            onChange={this.handleChange}
            spellCheck={false}
          />
          <Actions>
            <CreateButton
              type="submit"
              disabled={title === ""}
            >
              Save list
            </CreateButton>
            or
            <CancelButton
              onClick={() => this.toggleOpened()}
            >
              cancel
            </CancelButton>
          </Actions>
        </form>
      </Wrapper>
    ) : (
      <AddButton
        id={'column-add-btn'}
        onClick={() => this.toggleOpened()}
      >
        Add a new column...
      </AddButton>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addList: addList,
    }, dispatch)
  };
}


export default connect(null, mapDispatchToProps)(BoardAdder);
