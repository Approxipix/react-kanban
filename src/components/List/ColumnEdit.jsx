import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { editColumn } from '../../redux/rootReducer/actions';

const Form = styled.form`
  width: 90%;
`;

const Input = styled.input`
  padding: .55rem .5rem .5rem;
  width: 100%;
  border: none;
  font-weight: bold;
  box-shadow: none;
  font-size: 1rem;
  color: #40424b;
`;

class TaskEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      content: '',
    };
  }

  componentDidMount() {
    this.setState({
      title: this.props.list.title,
    });
    document.addEventListener('click', this.outerClick);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.outerClick);
  }

  outerClick = (e) => {
    let { target } = e;
    if (target.id === 'list-edit-form') {
      return;
    }
    if (!!target.closest && (target.closest('#list-edit-form'))) {
      return;
    }
    this.props.closeEdit();
  };

  handleChange = (key, value) => {
    this.setState({
      ...this.state,
      [key]: value,
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    const data = {
      boardIndex: this.props.boardIndex,
      listIndex: this.props.listIndex,
      listTitle: title,
    };
    this.props.actions.editColumn(data);
    this.props.closeEdit()
  };

  render = () => {
    const { title } = this.state;
    return (
      <Form onSubmit={this.handleSubmit} id="list-edit-form">
        <Input
          autoFocus
          type="text"
          placeholder="Board name"
          value={title}
          onChange={(e) => this.handleChange('title', e.target.value)}
          spellCheck={false}
        />
      </Form>
    );
  };
}

function mapStateToProps(state, ownProps) {
  return {
    list: state.rootReducer.boards[ownProps.boardIndex].list[ownProps.listIndex],
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editColumn: editColumn,
    }, dispatch)
  };
}


export default connect(mapStateToProps, mapDispatchToProps)(TaskEdit);
