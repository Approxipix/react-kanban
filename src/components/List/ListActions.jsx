import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { deleteList } from '../../redux/listReducer/actions';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome/index";
import styled from 'styled-components'
import ClickOutside from "../ClickOutside";

const Wrapper = styled.div`
  position: absolute;
  top: 10%;
  left: 60%;
  width: 10rem;
  background-color: #F4F5F7;
  border-radius: .2rem;
  box-shadow: rgba(0, 0, 0, 0.3) .125rem .125rem .5rem;
  z-index: 10;
`;

const Header = styled.div`
  position: relative;
  width: 100%;
  text-align: center;
  border-bottom: 1px solid rgba(9,30,66,.13);
`;

const Close = styled.button`
  position: absolute;
  top: .5rem;
  right: 0;
  color: #6b778c;
  transition: color .1s ease-in;
  cursor: pointer;
  &:hover {
    color: #42526e;
  }
`;

const Title = styled.h3`
  padding: .75rem .5rem;
  font-size: .9rem;
  color: #6b778c;
  font-weight: normal;
`;


const Action = styled.div`
  padding: .5rem 0;
`;

const Button = styled.button`
  width: 100%;
  padding: .5rem;
  font-size: .875rem;
  color: #40424b;
  font-weight: 600;
  text-align: left; 
  transition: all .2s ease-in;
  cursor: pointer;
  &:hover {
    color: #fff;
    background-color: #2E7EAF;
  }
`;

class ListActions extends Component {
  deleteList = () => {
    const { listId, boardId } = this.props;
    this.props.actions.deleteList({
      boardId: boardId,
      listId: listId,
    });
    this.props.toggleOpened();
  };

  render = () => {
    return (
      <ClickOutside toggleOpened={this.props.toggleOpened}>
        <Wrapper>
          <Header>
            <Title>
              List actions
            </Title>
            <Close onClick={() => this.props.toggleOpened()}>
              <FontAwesomeIcon icon="times" />
            </Close>
          </Header>
          <Action>
            <Button onClick={() => this.deleteList()}>
              Delete List
            </Button>
          </Action>
        </Wrapper>
      </ClickOutside>
    )
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      deleteList: deleteList,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(ListActions);
