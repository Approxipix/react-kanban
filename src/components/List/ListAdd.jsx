import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { addList } from '../../redux/listReducer/actions/actions';
import { Actions, Input, SubmitButton, CancelButton } from '../BaseComponent';
import ClickOutside from "../ClickOutside";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Wrapper = styled.div`
  flex-shrink:0;
  width: 16rem;
  margin: 0 .5rem;
`;

const Form = styled.form`
  padding: .6rem;
  background-color: #e3e3e3;
  border-radius: .2rem
`;

const AddButton = styled.button`
  width: 100%;
  padding: .6rem;
  font-size: 1rem; 
  color: #fff;
  text-align: left;
  border-radius: .2rem;
  background-color: rgba(0, 0, 0, .12);
  transition: background-color .2s ease-in;
  cursor: pointer;
  &:hover {
    background: rgba(0, 0, 0, 0.2);
  }
`;

class ListAdd extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
  }

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      title: '',
    })
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.toggleOpened();
    }
  };

  handleChange = (e) => {
    this.setState({
      title: e.target.value
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title) return;
    this.props.actions.addList({
      boardId: this.props.boardId,
      listTitle: title,
      listId: Date.now().toString(),
    });
    this.setState({
      title: ''
    }, () => this.input.focus())
  };

  render = () => {
    const { isOpened, title } = this.state;
    return isOpened ? (
      <ClickOutside toggleOpened={this.toggleOpened}>
        <Wrapper>
          <Form onSubmit={this.handleSubmit}>
            <Input
              type="text"
              value={title}
              placeholder="Enter a list title..."
              ref={(e) => { this.input = e }}
              onKeyDown={this.handleKeyDown}
              onChange={this.handleChange}
              spellCheck={false}
              autoFocus
            />
            <Actions>
              <SubmitButton type="submit" disabled={title === ""}>
                Add list
              </SubmitButton>
              <CancelButton onClick={() => this.toggleOpened()}>
                <FontAwesomeIcon icon="times"/>
              </CancelButton>
            </Actions>
          </Form>
        </Wrapper>
      </ClickOutside>
    ) : (
      <Wrapper>
        <AddButton onClick={() => this.toggleOpened()}>
          + Add another list
        </AddButton>
      </Wrapper>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addList: addList,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(ListAdd);
