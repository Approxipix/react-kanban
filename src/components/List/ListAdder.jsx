import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { addList } from '../../redux/boardReducer/actions';
import { Input, SubmitButton, CancelButton } from '../BaseComponent';
import ClickOutside from "../ClickOutside";

const Wrapper = styled.div`
  width: 16rem;
  margin: 0 .5rem;
  padding: .6rem;
  background-color: #e3e3e3;
  border-radius: .2rem
`;

const Title = styled.h4`
  margin-bottom: .5rem;
`;

const Form = styled.form``;


const Actions = styled.div`
  display: flex;
  align-items: center;
  font-size: .9rem;
`;

const AddButton = styled.button`
  width: 16rem;
  margin: 0 .5rem;
  padding: .75rem;
  border-radius: .2rem;
  background: rgba(0, 0, 0, 0.1);
  color: #c3c3c3;
  font-size: 1rem; 
  text-align: left;
  transition: background-color .1s;
  cursor: pointer;
  &:hover {
    background: rgba(0, 0, 0, 0.2);
  }
`;

class BoardAdder extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
  }

  handleChange = event => {
    this.setState({
      title: event.target.value
    });
  };

  handleKeyDown = event => {
    if (event.keyCode === 27) {
      this.setState({
        isOpened: false
      });
    }
  };

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened
    })
  };

  handleSubmit = event => {
    event.preventDefault();
    const { title } = this.state;
    if (!title) return;
    this.props.actions.addList({
      listTitle: title
    });
    this.toggleOpened();
  };

  render = () => {
    const { isOpened, title } = this.state;
    return isOpened ? (
      <ClickOutside toggleOpened={this.toggleOpened}>
        <Wrapper>
          <Title>
            New Column
          </Title>
          <Form id="column-add-form" onSubmit={this.handleSubmit}>
            <Input
              autoFocus
              type="text"
              placeholder="Enter a list title..."
              value={title}
              onKeyDown={this.handleKeyDown}
              onChange={this.handleChange}
              spellCheck={false}
            />
            <Actions>
              <SubmitButton type="submit" disabled={title === ""}>
                Add list
              </SubmitButton>
              or
              <CancelButton onClick={() => this.toggleOpened()}>
                cancel
              </CancelButton>
            </Actions>
          </Form>
        </Wrapper>
      </ClickOutside>
    ) : (
      <AddButton onClick={() => this.toggleOpened()}>
        + Add another list
      </AddButton>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addList: addList,
    }, dispatch)
  };
}


export default connect(null, mapDispatchToProps)(BoardAdder);
