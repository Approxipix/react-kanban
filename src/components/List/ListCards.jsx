import React, { Component, PureComponent } from 'react';
import styled from 'styled-components'
import { Droppable } from 'react-beautiful-dnd';
import Card from '../Card/Card'

const CardLIst = styled.div`
  min-height: 2rem;
  position: relative;
  padding: .5rem;
  flex-grow: 1;
  overflow: auto;
`;

class InnerList extends PureComponent {
  render() {
    const { listId, cardsId, } = this.props;
    return (
      cardsId.map((cardId, index) =>
        <Card
          isDraggingOver={this.props.isDraggingOver}
          key={index}
          index={index}
          listId={listId}
          cardId={cardId}
        />
      )
    )
  }
}

class ListCards extends Component {
  componentDidUpdate = prevProps => {
    const { cardsId } = this.props;
    if (
      cardsId[cardsId.length - 2] ===
      prevProps.cardsId[prevProps.cardsId.length - 1]
    ) {
      this.scrollToBottom();
    }
  };

  scrollToBottom = () => {
    this.listEnd.scrollIntoView();
  };

  render() {
    const { listId, cardsId } = this.props;
    return (
      <Droppable
        type={'task'}
        droppableId={`${listId}`}
      >
        {(provided, { isDraggingOver }) => (
          <CardLIst
            {...provided.droppableProps}
            ref={provided.innerRef}
            isDraggingOver={isDraggingOver}
          >
            <InnerList
              isDraggingOver={isDraggingOver}
              listId={listId}
              cardsId={cardsId}
            />
            {provided.placeholder}
            <div ref={el => this.listEnd = el}/>
          </CardLIst>
        )}
      </Droppable>
    )
  }
}

export default ListCards;
