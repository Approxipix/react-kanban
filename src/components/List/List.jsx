import React from 'react';
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Draggable } from 'react-beautiful-dnd';
import ListHeader from './ListHeader'
import CardAdd from '../Card/CardAdd';
import ListCards from './ListCards';
import styled from 'styled-components';

const Container = styled.div`
  position: relative;
  display: flex;
  flex-shrink: 0;
  flex-direction: column;
  max-height: 83vh;
  width: 16rem;
  margin: 0 .5rem;
  background-color: #dfe1e6;
  border-radius: .2rem;
`;

const List = (props) => {
  const {  boardId, list, index } = props;
  if (!list) return null;

  return (
    <Draggable index={index} draggableId={`${list._listId}`}>
      {(provided) => (
        <>
          <Container {...provided.draggableProps} ref={provided.innerRef}>
            <ListHeader
              dragHandleProps={provided.dragHandleProps}
              boardId={boardId}
              listId={list._listId}
              listTitle={list.title}
            />
            <ListCards listId={list._listId} cardsId={list.cards} />
            <CardAdd listId={list._listId} />
          </Container>
          {provided.placeholder}
        </>
      )}
    </Draggable>
  )
};

List.defaultProps = {
  boardId: '',
  listId: '',
  list: null
};

List.propTypes = {
  index: PropTypes.number,
  boardId: PropTypes.string.isRequired,
  listId: PropTypes.string.isRequired,
  list: PropTypes.shape({
    _listId: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    cards: PropTypes.arrayOf(PropTypes.string).isRequired
  }).isRequired,
};

function mapStateToProps(state, ownProps) {
  return {
    list: state.listReducer[ownProps.listId],
  }
}

export default connect(mapStateToProps)(List);
