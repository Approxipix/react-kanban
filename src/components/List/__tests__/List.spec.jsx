import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import List from '../List';

const mockStore = configureStore([]);
jest.mock('react-beautiful-dnd', () => ({
  Draggable: ({ children }) => children({
    draggableProps: {
      style: {},
    },
    innerRef: jest.fn(),
  }, {}),
}));

describe('<List>', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      listReducer: {
        ListID: {
          _listId: 'ListID',
          title: 'List Title',
          cards: []
        }
      },
    });
    const props = {
      listId: 'ListID'
    };

    const wrapper = shallow(<List store={store} {...props} />);
    component = wrapper.find('List').dive().dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
