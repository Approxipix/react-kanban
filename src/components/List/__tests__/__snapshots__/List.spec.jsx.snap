// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List> should render with given state from Redux store 1`] = `
<Fragment>
  <styled.div
    style={Object {}}
  >
    <ListHeader
      boardId=""
      listId="ListID"
      listTitle="List Title"
    />
    <ListCards
      cardsId={Array []}
      listId="ListID"
    />
    <Connect(CardAdd)
      listId="ListID"
    />
  </styled.div>
</Fragment>
`;
